lib/utils/token/token.go

```diff
@@ -0,0 +1,75 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package token
+
+import (
+	"math"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+// Verify ensures the token fits into our length requirements and
+// its bits of entropy are sufficient.
+func Verify(token []byte) error {
+	return verify(token, defaults.MaxTokenLength)
+}
+
+// VerifyHashed ensures the token fits into our length requirements and
+// its bits of entropy are sufficient. If the token is not going to be
+// hashed by bcrypt before it will be used, use [Verify] instead.
+func VerifyHashed(token []byte) error {
+	return verify(token, defaults.MaxHashedTokenLength)
+}
+
+func verify(token []byte, maxLen int) error {
+	if len(token) < defaults.MinTokenLength {
+		return trace.BadParameter("token is too short, min length is %d", defaults.MinTokenLength)
+	}
+	if len(token) > maxLen {
+		return trace.BadParameter("token is too long, max length is %d", maxLen)
+	}
+
+	entropyBits := EntropyBits(token)
+	if entropyBits < defaults.MinTokenEntropyBits {
+		return trace.BadParameter("token is not random enough, only has %d bits of entropy; min bits of entropy is %d", entropyBits, defaults.MinTokenEntropyBits)
+	}
+
+	return nil
+}
+
+// EntropyBits computes the number of bits of entropy the input has
+// using the formula E = log2(RL) where E is bits of entropy, R is
+// the range of different characters in the input, and L is the length
+// of the input.
```

Comment:
Is there a source for this algorithm to calculate an entropy value for a bytestring?

Reply:
I couldn't find a definitive source for the algorithm, it was mentioned pretty universally when searching for password entropy calculation algorithms.

Hmm that's a good point. Some examples I saw online used an assumed character range to estimate the input entropy, such as 62 if lowercase and uppercase letters were used as well as digits. Using the number of characters that appeared in the input seemed more useful to me though, as it represented the input better. Maybe we can combine this entropy estimation with the Shannon entropy calculation of the input to get a better estimate of the token's strength? I was originally just going to use Shannon entropy but it didn't seem to provide a good enough estimate to use by itself.

Reply:
What problems did you see with the standard shannon entropy?

Reply:
Shannon entropy doesn't take length into account, generally longer tokens are better as they're harder to guess, that was the main reason. I also wasn't thrilled that `abcdefghijklmnopqrstuvwxyz012345` has a higher entropy value than most hex-encoded CS random strings 16 bytes long (32 total bytes after encoding). The first string has no repeating characters so it has more entropy, but isn't a great token. I've been racking my brain but I'm not sure of a really good method of calculating token strength though. Right now the best I can come up with that's reasonably simple is `( L*log2(R) ) * shannon(input)`. How does that sound to you?

Reply:
Just do length times Shannon entropy? The Shannon entropy already takes the size of the alphabet into consideration, and it spits out the estimated entropy _per symbol_.